iscsi/iscsi.go

```diff
@@ -508,3 +511,83 @@ func GetConnectorFromFile(filePath string) (*Connector, error) {
 	return &data, nil
 
 }
+
+func RescanISCSIDevices(hctls []HCTL) error {
+	debug.Printf("Begin RescanISCSIDevices (%v)...", hctls)
+	for _, hctl := range hctls {
+		scanFilePath := fmt.Sprintf("/sys/class/scsi_host/host%d/scan", hctl.HBA)
+		err := os.WriteFile(scanFilePath, []byte(fmt.Sprintf("%d %d %d\n", hctl.Channel, hctl.Target, hctl.LUN)), 0644)
+		if err != nil {
+			debug.Printf("error writing scan file %s: %v", scanFilePath, err)
+			return err
+		}
+	}
+	return nil
+}
+
+// ISCSIRescan takes a target iqn and lun and writes to the scan file in the scsi subsystem
+// We do this manually instead of relying on iscsiadm -R. This prevents a race condition in which
+// devices that are in the process of being removed can be re-discovered and left behind.
+func ISCSIRescan(tgtIQN string, lun int) error {
+	debug.Printf("Begin ISCSIRescan (%s, %d)...", tgtIQN, lun)
+	var hctlsToScan []HCTL
+	// Get all scsi targets
+	sessionTargetFilenames, err := filepath.Glob("/sys/class/scsi_host/host*/device/session*/iscsi_session/session*/targetname")
+	if err != nil {
+		debug.Printf("Error searching for scsi session targets in /sys/class/scsi_host")
+		return err
+	}
+	SCSIHostPath := ""
+	// loop over all found sessions. if the targetname matches the target we want to scan, create an HCTL for it and add it to list of devices to scan
+	for _, sessionTargetFile := range sessionTargetFilenames {
+		targetName, err := os.ReadFile(sessionTargetFile)
+		if err != nil {
+			debug.Printf("Error reading session file %s, skipping to next session", sessionTargetFile)
+			continue
+		}
+		if strings.TrimSpace(string(targetName)) == strings.TrimSpace(tgtIQN) {
+			SCSIHostPath = strings.Split(sessionTargetFile, "/device/")[0]
+			hba, err := strconv.Atoi(strings.TrimPrefix(SCSIHostPath, "/sys/class/scsi_host/host"))
+			if err != nil {
+				debug.Printf("Error retrieving HBA number from path %s", SCSIHostPath)
+				return err
+			}
+			sessionPath := strings.Split(sessionTargetFile, "/iscsi_session")[0]
+			targetFilesInSession, err := filepath.Glob(filepath.Join(sessionPath, "target*"))
+			if err != nil {
+				debug.Printf("Error getting target info from session directory %s", sessionPath)
+				return err
+			}
+			for _, target := range targetFilesInSession {
+				// this will be a filename formatted like "target3:0:0", we want to extract the last 2 numbers which represent the channel and target
+				hostChannelTarget := strings.Split(strings.TrimPrefix(filepath.Base(target), "target"), ":")
+				channel, err := strconv.Atoi(hostChannelTarget[1])
```

Comment:
Here's hoping len(hostChannelTarget) >= 3

Reply:
Added length check for hostChannelTarget slice